app/code/Magento/Catalog/Model/ResourceModel/Category.php

```diff
@@ -113,8 +128,21 @@ public function __construct(
         $this->_categoryCollectionFactory = $categoryCollectionFactory;
         $this->_eventManager = $eventManager;
         $this->connectionName  = 'catalog';
-        $this->serializer = $serializer ?: \Magento\Framework\App\ObjectManager::getInstance()
+        $this->serializer = $serializer ?: ObjectManager::getInstance()
             ->get(\Magento\Framework\Serialize\Serializer\Json::class);
+        $this->metadataService = $metaDataService ?? ObjectManager::getInstance()
+            ->get(CategoryAttributeRepositoryInterface::class);
+    }
+
+    public function getCustomAttributesCodes()
+    {
+        if ($this->customAttributesCodes === null) {
+            $this->customAttributesCodes = $this->getEavAttributesCodes($this->metadataService);
+            $this->customAttributesCodes = array_values(
+                array_diff($this->customAttributesCodes, CategoryInterface::ATTRIBUTES)
+            );
+        }
+        return $this->customAttributesCodes;
```

Comment:
In the case of products we can improve this even further for larger datasets, as it stands here it will work great for a small number of attributes (< 500) but as soon as we start to go upward of that figure the performance will fall through the floor (each attribute loaded results in `Magento\Eav\Model\ResourceModel\Entity\Attribute::_afterLoad` being called which in turn runs a select query to join additional attribute data).

In my current project we have ~2200 custom attributes, but all of our attribute sets are small (< 100 custom attributes in each one), the problem is that this method loads every attribute regardless of the product's attribute set, to fix it I pulled in this PR as a patch and then changed this method to the following:

```php
    public function getCustomAttributesCodes($attributeSetId)
    {
        if (!isset($this->customAttributesCodes[$attributeSetId])) {
            $codes = [];
            $criteria = $this->searchCriteriaBuilder->addFilter('attribute_set_id', $attributeSetId, 'eq');
            $attributes = $this->attributeRepository->getList($criteria->create())->getItems();

            if (is_array($attributes)) {
                /** @var $attribute \Magento\Framework\Api\MetadataObjectInterface */
                foreach ($attributes as $attribute) {
                    $codes[] = $attribute->getAttributeCode();
                }
            }

            $codes = array_values(
                array_diff($codes, ProductInterface::ATTRIBUTES)
            );

            $this->customAttributesCodes[$attributeSetId] = $codes;
        }

        return $this->customAttributesCodes[$attributeSetId];
    }
```

then had the `Magento\Catalog\Model\Product` pass through the product's attribute set ID, this way we only ever load the attributes for the current attribute set

I think the memoization based on attribute set ID in the `$this->customAttributesCodes[$attributeSetId]` is not required because resource model instances are not shared across products, but I might be wrong  so I've left it in there for safety

whilst this would reduce in a few more queries (1 for each different attribute set to get the attributes relevant for that attribute) it means that creating a product is not affected by the total number of attributes in Magento, only by the number of attributes in the product's attribute set

Reply:
> because resource model instances are not shared across products, 

They are, this is what my initial approach is based on. Thanks a lot for your addition, it makes sense and I'd add it to the PR!

But first waiting for @maghamed to discuss the final architecture